Sources/SIL/SExpr.swift

```diff
@@ -0,0 +1,126 @@
+// This is a temporary addition to libSIL and will likely be removed in the future.
+// It is a hack that allows one to read the output of e.g. swiftc -dump-ast.
+public enum SExpr : Equatable {
+    public enum Property : Equatable {
+      case value(SExpr)
+      case field(String, SExpr)
+    }
+
+    case symbol(String)
+    case string(String)
+    case sourceRange(String) // We don't parse those further right now
+    case record(String, [Property])
+
+    public static func parse(fromPath path: String) throws -> SExpr {
+        let parser = try SExprParser(forPath: path)
+        return try parser.parseExpr()
+    }
+}
+
+class SExprParser: Parser {
+    func parseExpr() throws -> SExpr {
+        if skip("(") {
+            var properties: [SExpr.Property] = []
+            guard case let .symbol(name) = try parseExpr() else {
+                throw parseError("Expected an expression body to start with a symbol")
+            }
+            while !skip(")") {
+                let expr = try parseExpr()
+                if case let .symbol(propName) = expr, skip("=") {
+                    properties.append(.field(propName, try parseValue()))
+                } else {
+                    if case let .symbol(exprValue) = expr {
+                        guard !exprValue.isEmpty else {
+                            throw parseError("An empty property?")
+                        }
+                    }
+                    properties.append(.value(expr))
+                }
+            }
+            return .record(name, properties)
+        }
+        return try parseValue()
+    }
+
+    func parseValue() throws -> SExpr {
+        if skip("'") {
+            let result = take(while: { $0 != "'" })
+            try take("'")
+            return .string(result)
+        }
+        if skip("\"") {
+            let result = take(while: { $0 != "\"" })
+            try take("\"")
+            return .string(result)
+        }
+        if skip("[") {
+            let result = take(while: { $0 != "]" })
+            try take("]")
+            return .sourceRange(result)
+        }
+        return try parseSymbol()
+    }
+
+    func parseSymbol() throws -> SExpr {
+        // NB: This is more complicated than it ever should be because swiftc
+        //     likes to print badly formed symbols that look like Module.(file).Type
+        var result = ""
+        while true {
+            let c = peek()
+            if !c.isWhitespace && !"()=".contains(c) {
+                let cs = String(c)
+                try take(cs, keepingTrivia: true)
+                result += cs
+                continue
+            }
+            if c == "(", skip("(file)") {
+                result += "(file)"
+                continue
+            }
+            break
+        }
+        skipTrivia()
```

Comment:
It cannot. Parenthesis can appear inside symbols because of `(file)`, so the parsing rules cannot be implemented without pretty complex lookaheads.

Reply:
How about `take(while: fn)` where `fn` counts the balance of opening/closing parentheses? I'm not a fan of the fact that we need three new parser APIs (`peek`, `keepingTrivia` and `skipTrivia` to handle this), so it would be good to find an alternative.

Reply:
But do we really want to keep track of those? This seems to complicate the context-free properties of `take` which I'd argue are desirable, plus I really don't want to allow anything else than this weird `(file)` component to appear in the symbol. Otherwise we would accept e.g. nested trees that are not separated using a space as if nothing has happened.

Reply:
Sorry for a cryptic comment - I was on the run. I didn't mean to change `take`, I meant to write a `fn` that would keep track of this stuff. If you'd like, this function could also only allow `(file)`. I'd just really want to avoid expanding the API of Parser.

Reply:
You'd really want to avoid expanding the API of Parser at a cost of adding highly questionable and specialized methods to a generic class? I probably still don't fully understand what you want to do.